looplib/loopstats.py

```diff
@@ -0,0 +1,170 @@
+#            boundary_list (Kx1 array of monomer positions)
+
+
+def calc_LEF_stalling_by_leg(LEF_array, polymer_length, boundary_list):
+    ''' calculates the fraction of LEFs with two, one, or no legs stalled at boundaries.
+        returns the fraction with both legs, one leg, or neither leg overlapping boundary_list    '''
+    boundary_list = np.array(boundary_list,dtype=int).flatten()
+    isBoundary =  np.histogram(boundary_list,  np.arange(0,polymer_length+1))[0]
+    LEF_arm_status = np.sum( isBoundary[LEF_array] , 1)
+    return  [(np.sum(LEF_arm_status==2)/len(LEF_array) ),  (np.sum(LEF_arm_status==1)/len(LEF_array) ) ,(np.sum(LEF_arm_status==0)/len(LEF_array)  )]
+stat_output_dict[ calc_LEF_stalling_by_leg] = ['both_stalled','one_stalled','none_stalled']
+
+def calc_boundary_occupancy(LEF_array, polymer_length, boundary_list):
+    ''' calculates the fraction of LEFs with two, one, or no legs stalled at boundaries.
+        returns the fraction with both legs, one leg, or neither leg overlapping boundary_list  '''
+    boundary_list = np.array(boundary_list,dtype=int).flatten()
+    bb = np.arange(0,polymer_length+1, 1); bb_mids = .5*(bb[:-1]+bb[1:])
+    extruderHistogram, b = np.histogram( LEF_array ,bb)
+    boundary_occupancy = np.mean( extruderHistogram[boundary_list ] )
+    non_boundary_list = np.setdiff1d(np.arange(1,polymer_length-1), boundary_list)
+    non_occupancy = np.mean( extruderHistogram[non_boundary_list ]  )
+    return [boundary_occupancy, non_occupancy]
+stat_output_dict[ calc_boundary_occupancy] = ['boundary_occupancy','non_occupancy']
+
+
+def calc_boundary_crossing_percent(LEF_array, polymer_length, boundary_list):
+    ''' calculates the fraction of LEFs with two, one, or no legs stalled at boundaries.
+        returns the fraction with both legs, one leg, or neither leg overlapping boundary_list  '''
+    boundary_list = np.array(boundary_list,dtype=int).flatten()
+    allLocs_df = pd.DataFrame(boundary_list, columns=['start'])
+    allLocs_df['end'] = boundary_list
+    allLocs_df['chr'] = 13
+    allLocs_df = allLocs_df[['chr','start','end']]
+    
+    LEF_array_df = pd.DataFrame(LEF_array,columns=['start','end'])
+    LEF_array_df['chr'] = 13
+    LEF_array_df= LEF_array_df[['chr','start','end']]
+    LEF_array_df.sort_values(['start'], inplace=True)                    
+    LEF_array_shrink_df = LEF_array_df.copy()
+    LEF_array_shrink_df['start'] = LEF_array_shrink_df['start'].values+1
+    LEF_array_shrink_df['end']   = np.maximum( LEF_array_shrink_df['end'].values-1, LEF_array_shrink_df['start'])
+                    
+    percent_crossing = np.sum((bedtools_intersect_counts( LEF_array_shrink_df, allLocs_df)['counts'] > 0)  
+                            / len(LEF_array) )
+    return [percent_crossing]
+stat_output_dict[ calc_boundary_crossing_percent] = ['percent_crossing']
+
+def calc_numLoops_given_boundaryPair_contact(LEF_array, polymer_length, boundary_pair_array, data):
```

Comment:
this depends on xyz coordinate of polymer (`data`) so probably shouldn't be in looplib...